From d7a9e17a242f5aff317b8b985d487367452ee5dc Mon Sep 17 00:00:00 2001
From: Merlin Kooshmanian <mkooshmanian@gmail.com>
Date: Wed, 29 Oct 2025 13:49:21 +0100
Subject: [PATCH 05/19] sched/deadline: Account TG bandwidth servers in 
 schedulability tests

Account the TG bandwidth server hierarchy in the schedulability checks
performed for deadline entities and global bandwidth updates. Integrate
the root task group's reserved TG bandwidth into overflow detection and
sched_dl_global_validate() so we honor the reservation when updating
runtime settings.

Largely inspired by the HCBS patch set developed by Luca Abeni and collaborators.

Based-on-a-patch-by: Luca Abeni <luca.abeni@santannapisa.it>
Co-developed-by: Alessio Balsini <a.balsini@sssup.it>
Co-developed-by: Andrea Parri <parri.andrea@gmail.com>
Co-developed-by: Yuri Andriaccio <yurand2000@gmail.com>
---
 kernel/sched/deadline.c | 21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/kernel/sched/deadline.c b/kernel/sched/deadline.c
index 2efb42030d07..bd8601f57ec7 100644
--- a/kernel/sched/deadline.c
+++ b/kernel/sched/deadline.c
@@ -212,8 +212,15 @@ void __dl_add(struct dl_bw *dl_b, u64 tsk_bw, int cpus)
 static inline bool
 __dl_overflow(struct dl_bw *dl_b, unsigned long cap, u64 old_bw, u64 new_bw)
 {
+	u64 tg_servers_root = 0;
+
+#ifdef CONFIG_TG_BANDWIDTH_SERVER
+	tg_servers_root = to_ratio(root_task_group.tg_bandwidth.dl_period,
+				  root_task_group.tg_bandwidth.dl_runtime);
+#endif
 	return dl_b->bw != -1 &&
-	       cap_scale(dl_b->bw, cap) < dl_b->total_bw - old_bw + new_bw;
+	       cap_scale(dl_b->bw, cap) < dl_b->total_bw - old_bw + new_bw
+					+ cap_scale(tg_servers_root, cap);
 }
 
 static inline
@@ -3110,10 +3117,16 @@ int sched_dl_global_validate(void)
 	u64 period = global_rt_period();
 	u64 new_bw = to_ratio(period, runtime);
 	u64 cookie = ++dl_cookie;
+	u64 tg_servers_root = 0;
 	struct dl_bw *dl_b;
-	int cpu, cpus, ret = 0;
+	int cpu, cap, cpus, ret = 0;
 	unsigned long flags;
 
+#ifdef CONFIG_TG_BANDWIDTH_SERVER
+	tg_servers_root = to_ratio(root_task_group.tg_bandwidth.dl_period,
+				  root_task_group.tg_bandwidth.dl_runtime);
+#endif
+
 	/*
 	 * Here we want to check the bandwidth not being set to some
 	 * value smaller than the currently allocated bandwidth in
@@ -3126,10 +3139,12 @@ int sched_dl_global_validate(void)
 			goto next;
 
 		dl_b = dl_bw_of(cpu);
+		cap = dl_bw_capacity(cpu);
 		cpus = dl_bw_cpus(cpu);
 
 		raw_spin_lock_irqsave(&dl_b->lock, flags);
-		if (new_bw * cpus < dl_b->total_bw)
+		if (new_bw * cpus < dl_b->total_bw +
+				cap_scale(tg_servers_root, cap))
 			ret = -EBUSY;
 		raw_spin_unlock_irqrestore(&dl_b->lock, flags);
 
-- 
2.43.0

